### PR TITLE
lightning: support base64 encoding of password

### PR DESCRIPTION
**Pull request by gozssky**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/31194

Problem Summary:

Let tidb-lightning supports base64 encoding of password.

### What is changed and how it works?

Let lightning `--tidb-password` in the command line and `password` under `tidb` section compatible with base64 encoding of password.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
lightning: compatible with base64 password
```

### CODE REVIEW

File: br/pkg/lightning/common/util.go [hunk 1]
```diff
@@ -64,15 +68,42 @@ func (param *MySQLConnectParam) ToDSN() string {
...
 
+func (param *MySQLConnectParam) Connect() (*sql.DB, error) {
+	db, err := param.tryConnect(param.Password)
+	if err == nil {
+		return db, nil
+	}
+	// If access is denied and password is encoded by base64, try the decoded string as well.
+	if mysqlErr, ok := errors.Cause(err).(*mysql.MySQLError); ok && mysqlErr.Number == tmysql.ErrAccessDenied {
+		// If password is encoded by base64, try the decoded string as well.
+		if password, decodeErr := base64.StdEncoding.DecodeString(param.Password); decodeErr == nil {
```
Comment (glorv) [inline br/pkg/lightning/common/util.go:96]: How about try base decode first, if success, we try decoded password to connect first and if failed, then try the origin one.

**Review by glorv — COMMENTED** [review]
rest LGTM

File: br/pkg/lightning/common/util.go (hunk 1 above)
Comment (gozssky) [reply br/pkg/lightning/common/util.go:96]: If password only contains digits, it always can be decoded as base64 string. So to keep consistent with previous behavior and avoid unsuccessfully database connect, here always tries plain password first.

**Review by glorv — COMMENTED** [review]
LGTM

**Review by glorv — APPROVED** [review]

**Comment by gozssky** [comment]
@kennytm PTAL, thanks.

**Review by lichunzhu — APPROVED** [review]
LGTM, BTW do we need a simple encrypt like dm?
https://github.com/pingcap/tiflow/blob/a4d10769ed929de10891e4e21c5be81525341805/dm/pkg/encrypt/encrypt.go#L67

**Comment by gozssky** [comment]
> LGTM, BTW do we need a simple encrypt like dm? https://github.com/pingcap/tiflow/blob/a4d10769ed929de10891e4e21c5be81525341805/dm/pkg/encrypt/encrypt.go#L67

@lichunzhu Thanks for your review! Lightning usually runs as a standalone binary. There is no place to save the secret key safely. That says secret key is equivalent with the plain password which I think encryption is a bit meaningless. However, If we fully integrate Lightning into other service (e.g DM) and deprecate binary in the future, password encryption can be reconsidered then.

**Comment by lichunzhu** [comment]
> > LGTM, BTW do we need a simple encrypt like dm? https://github.com/pingcap/tiflow/blob/a4d10769ed929de10891e4e21c5be81525341805/dm/pkg/encrypt/encrypt.go#L67
> 
> @lichunzhu Thanks for your review! Lightning usually runs as a standalone binary. There is no place to save the secret key safely. That says secret key is equivalent with the plain password which I think encryption is a bit meaningless. However, If we fully integrate Lightning into other service (e.g DM) and deprecate the binary release in the future, password encryption can be reconsidered then.

SGTM

**Comment by gozssky** [comment]
@lichunzhu Could you help merge this PR?

**Pull request merged**